File: 백준/Gold/16974. 레벨 햄버거/레벨 햄버거.py
import sys
read = sys.stdin.readline

n, x = map(int, read().split())
burgers = [0] * 51
patties = [0] * 51
burgers[0] = 1
patties[0] = 1
for i in range(1, 51):
    burgers[i] = 3 + 2 * burgers[i-1]
    patties[i] = 1 + 2 * patties[i-1]

# 반씩 쪼개기
ans = 0
curL = n
target = x-1

while curL >= 0:
    if target == burgers[curL] - 1:
        ans += patties[curL]
        break
    if target == 0 and curL != 0:
        break   
    if curL == 0:
        ans += 1
        break

    mid = (burgers[curL]) // 2
    if target == mid:
        ans += patties[curL-1] + 1
        break
    elif target > mid:
        ans += patties[curL-1] + 1
        target = target - mid - 1
        curL -= 1
    else: # target < mid
        target -= 1
        curL -= 1

print(ans)

File: 백준/Gold/16974. 레벨 햄버거/레벨 햄버거.js
const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "sample.txt")
  .toString()
  .trim()
  .split("\n");

const [n, x] = input[0].split(" ").map(Number);
solve(n, x);


function solve(n, x) {
  const burgers = Array(n + 1).fill(0); // 각 레이어별 패티 개수
  let size = 1n; // 버거 크기
  burgers[0] = 1n;
  for (let i = 1; i < n + 1; i++) {
    burgers[i] = 1n + 2n * burgers[i - 1];
    size = 3n + 2n * size;
  }
  let [l, r] = [1n, BigInt(size)];
  let answer = 0n;
  // r에서부터 x개를 체크하면서 분할 탐색
  x = BigInt(x);
  while (l <= r) {
    let xPos = r - x + 1n; // r에서 x개 먹을 때 위치
    // l, r 사이를 벗어났을때
    if (xPos < l) {
      answer += burgers[n];
      break;
    }
    if (xPos > r) {
      break;
    }
    let mid = (l + r) / 2n;
    if (xPos === mid) {
      // 중간
      answer += 1n + (n > 0 ? burgers[n - 1] : 0n);
      break;
    } else if (xPos < mid) {
      // 중간의 왼쪽
      answer += 1n + burgers[n - 1];
      x -= r - mid + 1n
      r = mid - 1n;
      l++;
      n--;
    } else {
      // 중간의 오른쪽a
      x--;
      r--;
      l = mid + 1n;
      n--;
    }
  }
  console.log(String(answer));

}
